// @mdingena/tiny-typed-emitter/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypedEmitter = void 0;
const node_events_1 = require("node:events");
class TypedEmitter extends node_events_1.EventEmitter {
    static defaultMaxListeners;
    addListener(event, listener) {
        super.addListener(typeof event === 'number' ? String(event) : event, listener);
        return this;
    }
    prependListener(event, listener) {
        super.prependListener(event, listener);
        return this;
    }
    prependOnceListener(event, listener) {
        super.prependOnceListener(event, listener);
        return this;
    }
    removeListener(event, listener) {
        super.removeListener(event, listener);
        return this;
    }
    removeAllListeners(event) {
        super.removeAllListeners(event);
        return this;
    }
    once(event, listener) {
        super.once(event, listener);
        return this;
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
    off(event, listener) {
        super.off(event, listener);
        return this;
    }
    emit(event, ...args) {
        return super.emit(event, ...args);
    }
    eventNames() {
        return super.eventNames();
    }
    listenerCount(type) {
        return super.listenerCount(type);
    }
    listeners(type) {
        return super.listeners(type);
    }
    rawListeners(type) {
        return super.rawListeners(type);
    }
    getMaxListeners() {
        return super.getMaxListeners();
    }
    setMaxListeners(n) {
        super.setMaxListeners(n);
        return this;
    }
}
exports.TypedEmitter = TypedEmitter;
