// att-client/dist/cjs/SubscriptionsManager/SubscriptionsManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionsManager = void 0;
const Subscriptions_js_1 = require("../Subscriptions/Subscriptions.js");
class SubscriptionsManager {
    client;
    instances;
    nextInstanceId;
    subscriptionsMap;
    constructor(client) {
        this.client = client;
        this.instances = new Map();
        this.nextInstanceId = 0;
        this.subscriptionsMap = new Map();
    }
    async subscribe(event, key, callback) {
        const subscription = `${event}/${key}`;
        if (this.subscriptionsMap.has(subscription)) {
            this.client.logger.error(`[SUBSCRIPTIONS-MANAGER] Already subscribed to ${subscription}.`);
            return;
        }
        const [instanceId, subscriptions] = await this.getInstanceWithCapacity();
        const subscribeResult = await subscriptions.subscribe(event, key, callback);
        this.subscriptionsMap.set(subscription, instanceId);
        return subscribeResult;
    }
    async unsubscribe(event, key) {
        const subscription = `${event}/${key}`;
        const instanceId = this.subscriptionsMap.get(subscription);
        if (typeof instanceId === 'undefined') {
            this.client.logger.error(`[SUBSCRIPTIONS-MANAGER] Subscription to ${subscription} does not exist.`);
            return;
        }
        const subscriptions = this.instances.get(instanceId);
        if (typeof subscriptions === 'undefined') {
            this.client.logger.error(`[SUBSCRIPTIONS-MANAGER] Couldn't retrieve managed subscriptions instance matching subscription to ${subscription}.`);
            return;
        }
        const unsubscribeResult = await subscriptions.unsubscribe(event, key);
        this.subscriptionsMap.delete(subscription);
        if (subscriptions.getSize() === 0) {
            this.client.logger.debug(`[SUBSCRIPTIONS-MANAGER] Subscriptions instance ${instanceId} is empty. Cleaning up.`);
            this.instances.delete(instanceId);
        }
        return unsubscribeResult;
    }
    getInstanceId() {
        return this.nextInstanceId++;
    }
    async getInstanceWithCapacity() {
        const maxSubscriptions = this.client.config.maxSubscriptionsPerWebSocket;
        for (const [instanceId, subscriptions] of this.instances) {
            if (subscriptions.getSize() < maxSubscriptions)
                return [instanceId, subscriptions];
        }
        this.client.logger.debug(`[SUBSCRIPTIONS-MANAGER] Increasing the Subscriptions instance pool.`);
        const instanceId = this.getInstanceId();
        const subscriptions = new Subscriptions_js_1.Subscriptions(this.client, instanceId);
        this.instances.set(instanceId, subscriptions);
        await subscriptions.init();
        this.client.logger.debug(`[SUBSCRIPTIONS-MANAGER] Created new Subscriptions instance with ID ${instanceId}.`);
        return [instanceId, subscriptions];
    }
}
exports.SubscriptionsManager = SubscriptionsManager;
