// att-client/dist/cjs/Logger/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const nativeConsole = console;
const noOpFn = (...args) => void 0;
class Logger {
    logError;
    logWarn;
    logInfo;
    logDebug;
    error;
    warn;
    info;
    debug;
    constructor({ verbosity, console = nativeConsole, prefix }) {
        this.logError = verbosity >= 1 ? console.error : noOpFn;
        this.logWarn = verbosity >= 2 ? console.warn : noOpFn;
        this.logInfo = verbosity >= 3 ? console.info : noOpFn;
        this.logDebug = verbosity >= 4 ? console.debug : noOpFn;
        const separator = prefix.length === 0 ? '' : ' ';
        this.error = (...args) => this.logError(`${prefix}${separator}error ${args.shift()}`, ...args);
        this.warn = (...args) => this.logWarn(`${prefix}${separator}warn ${args.shift()}`, ...args);
        this.info = (...args) => this.logInfo(`${prefix}${separator}info ${args.shift()}`, ...args);
        this.debug = (...args) => this.logDebug(`${prefix}${separator}debug ${args.shift()}`, ...args);
    }
}
exports.Logger = Logger;
