// att-client/dist/cjs/Workers/Workers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workers = void 0;
class Workers {
    maxConcurrency;
    pool;
    tasks;
    workers;
    constructor(maxConcurrency) {
        this.maxConcurrency = maxConcurrency;
        this.pool = new Map();
        this.tasks = [].values();
        this.workers = new Array(maxConcurrency);
    }
    async do(tasks, callback) {
        this.tasks = tasks.values();
        for (let i = 0; i < this.maxConcurrency; i++) {
            this.workers[i] = this.createWorker();
        }
        this.pool = new Map(this.workers.map(worker => [worker, this.assign(worker)]));
        const results = [];
        for await (const taskResult of this.watch()) {
            results.push(taskResult);
            callback?.(taskResult);
        }
        return results;
    }
    async *createWorker() {
        for (const task of this.tasks) {
            yield await task();
        }
    }
    async assign(worker) {
        const task = worker.next();
        return { worker, result: await task };
    }
    async *watch() {
        while (this.pool.size) {
            const pooledWork = this.pool.values();
            const completedWork = await Promise.race(pooledWork);
            const { worker, result: { value, done } } = completedWork;
            if (done) {
                this.pool.delete(worker);
            }
            else {
                this.pool.set(worker, this.assign(worker));
                yield value;
            }
        }
    }
}
exports.Workers = Workers;
