// att-client/dist/esm/Group/Group.js
import { TypedEmitter } from '@mdingena/tiny-typed-emitter';
import { Server } from '../Server/index.js';
export class Group extends TypedEmitter {
    client;
    description;
    id;
    name;
    permissions;
    roles;
    servers;
    userId;
    constructor(client, group, member) {
        super();
        this.client = client;
        this.description = group.description ?? '';
        this.id = group.id;
        this.name = group.name ?? '';
        this.permissions = this.getPermissions(group, member);
        this.roles = [];
        this.servers = {};
        this.userId = member.user_id;
        if (!this.permissions.includes('Console')) {
            this.client.logger.warn(`[GROUP-${this.id}] Client does not have 'Console' permissions.`, JSON.stringify(this.permissions));
        }
        this.addServers(group);
    }
    async init() {
        await Promise.all([
            this.client.subscriptions.subscribe('group-update', this.id.toString(), async (message) => {
                try {
                    const group = message.content;
                    this.updateGroup(group);
                }
                catch (error) {
                    this.client.logger.error(`[GROUP-${this.id}] Error while handling group update: ${error.message}`);
                }
            }),
            this.client.subscriptions.subscribe('group-member-update', this.id.toString(), async (message) => {
                try {
                    const member = message.content;
                    if (member.user_id !== this.userId)
                        return;
                    this.client.logger.info(`[GROUP-${this.id}] Membership updated.`);
                    const group = await this.client.api.getGroupInfo(this.id);
                    if (typeof group === 'undefined') {
                        this.client.logger.error(`[GROUP-${this.id}] Couldn't get group info.`);
                        return;
                    }
                    this.updateGroup(group, member);
                }
                catch (error) {
                    this.client.logger.error(`Error while handling group member update: ${error.message}`);
                }
            }),
            this.client.subscriptions.subscribe('group-server-status', this.id.toString(), async (message) => {
                try {
                    const status = message.content;
                    this.client.logger.debug(`[GROUP-${this.id}] Status updated for server ${status.id} (${status.name}).`, JSON.stringify(status));
                    this.manageServerConnection(status);
                    const server = this.servers[status.id];
                    if (typeof server !== 'undefined') {
                        server.update(status);
                    }
                }
                catch (error) {
                    this.client.logger.error(`[GROUP-${this.id}] Error while handling server update: ${error.message}`);
                }
            }),
            this.client.subscriptions.subscribe('group-server-heartbeat', this.id.toString(), async (message) => {
                try {
                    const status = message.content;
                    this.client.logger.debug(`[GROUP-${this.id}] Heartbeat for server ${status.id} (${status.name}).`, JSON.stringify(status));
                    this.handleHeartbeat(status);
                }
                catch (error) {
                    this.client.logger.error(`[GROUP-${this.id}] Error while handling server heartbeat: ${error.message}`);
                }
            }),
            this.client.subscriptions.subscribe('group-server-create', this.id.toString(), _unstableMessage => {
                try {
                    this.client.logger.warn(`[GROUP-${this.id}] Client is running untested group-server-create code in Group.ts.`, JSON.stringify(_unstableMessage));
                    const serverId = _unstableMessage.content.id;
                    this.addServer(serverId);
                }
                catch (error) {
                    this.client.logger.error(`[GROUP-${this.id}] Error while handling server creation: ${error.message}`);
                }
            }),
            this.client.subscriptions.subscribe('group-server-delete', this.id.toString(), _unstableMessage => {
                try {
                    this.client.logger.warn(`[GROUP-${this.id}] Client is running untested group-server-delete code in Group.ts.`, JSON.stringify(_unstableMessage));
                    const serverId = _unstableMessage.content.id;
                    this.removeServer(serverId);
                }
                catch (error) {
                    this.client.logger.error(`[GROUP-${this.id}] Error while handling server deletion: ${error.message}`);
                }
            })
        ]);
        await this.updateServers();
    }
    async updateServers() {
        return await Promise.all(Object.values(this.servers).map(server => this.updateServer(server.id)));
    }
    async updateServer(serverId) {
        this.client.logger.debug(`[GROUP-${this.id}] Updating info for server ${serverId}.`);
        const status = await this.client.api.getServerInfo(serverId);
        if (typeof status === 'undefined') {
            this.client.logger.error(`[GROUP-${this.id}] Couldn't get status for server ${serverId} (${this.name}).`);
            return;
        }
        this.manageServerConnection(status);
    }
    async updateGroup(group, member) {
        this.name = group.name ?? this.name;
        this.description = group.description ?? this.description;
        this.roles =
            group.roles?.map(role => ({ id: role.role_id, name: role.name ?? '', permissions: role.permissions })) ??
                this.roles;
        if (typeof member !== 'undefined') {
            await this.updatePermissions(group, member);
        }
        this.emit('update', this);
    }
    async updatePermissions(group, member) {
        const previousPermissions = [...this.permissions];
        this.permissions = this.getPermissions(group, member);
        if (!previousPermissions.includes('Console') && this.permissions.includes('Console')) {
            this.client.logger.info(`[GROUP-${this.id}] Client gained console access to servers in group.`);
        }
        else if (previousPermissions.includes('Console') && !this.permissions.includes('Console')) {
            this.client.logger.info(`[GROUP-${this.id}] Client lost console access to servers in group.`);
        }
        await this.updateServers();
    }
    getPermissions(group, member) {
        const roleId = member.role_id;
        const roles = group.roles ?? [];
        const role = roles.find(role => role.role_id === roleId);
        return role?.permissions ?? [];
    }
    async handleHeartbeat(status) {
        const serverId = status.id;
        const server = await this.ensureServer(serverId);
        if (typeof server === 'undefined') {
            this.client.logger.error(`[GROUP-${this.id}] Server ${serverId} not found in group.`);
            return;
        }
        server.handleHeartbeat(status);
        this.manageServerConnection(status);
    }
    async manageServerConnection(status) {
        const serverId = status.id;
        const server = await this.ensureServer(serverId);
        if (typeof server === 'undefined') {
            this.client.logger.error(`[GROUP-${this.id}] Server ${serverId} not found in group.`);
            return;
        }
        const hasConsolePermission = this.permissions.includes('Console');
        const isSupportedServerFleet = this.client.config.supportedServerFleets.includes(server.fleet);
        const mayConnect = hasConsolePermission && isSupportedServerFleet;
        const isServerOnline = status.is_online;
        const hasOnlinePlayers = status.online_players.length > 0;
        if (server.status === 'disconnected' && mayConnect && isServerOnline && hasOnlinePlayers) {
            try {
                await server.connect();
            }
            catch (error) {
                this.client.logger.error(`[GROUP-${this.id}] Couldn't connect to server ${serverId}: ${error.message}`);
            }
        }
        else if (server.status !== 'disconnected' && (!mayConnect || !isServerOnline)) {
            server.disconnect();
        }
    }
    addServers(group) {
        this.client.logger.debug(`[GROUP-${this.id}] Adding all servers for group.`);
        for (const server of group.servers ?? []) {
            this.addServer(server.id);
        }
    }
    async addServer(serverId) {
        this.client.logger.info(`[GROUP-${this.id}] Managing server ${serverId} (${this.name}).`);
        try {
            if (Object.keys(this.servers).map(Number).includes(serverId)) {
                throw new Error(`[GROUP-${this.id}] Can't add server ${serverId} (${this.name}) more than once.`);
            }
            const server = await this.client.api.getServerInfo(serverId);
            const managedServer = new Server(this, server);
            this.servers = {
                ...this.servers,
                [serverId]: managedServer
            };
            this.emit('server-add', managedServer);
            this.manageServerConnection(server);
            return managedServer;
        }
        catch (error) {
            this.client.logger.error(`[GROUP-${this.id}] Couldn't add server ${serverId} to group: ${error.message}`);
        }
        return;
    }
    removeServers() {
        this.client.logger.debug(`[GROUP-${this.id}] Removing all servers from group.`);
        for (const server of Object.values(this.servers)) {
            this.removeServer(server.id);
        }
    }
    removeServer(serverId) {
        this.client.logger.info(`[GROUP-${this.id}] Removing server ${serverId} (${this.name}).`);
        const server = this.servers[serverId];
        if (typeof server === 'undefined') {
            this.client.logger.error(`[GROUP-${this.id}] Can't remove an unmanaged server with ID ${serverId} (${this.name}).`);
            return;
        }
        server.dispose();
        delete this.servers[serverId];
    }
    async ensureServer(serverId) {
        let server = this.servers[serverId];
        if (typeof server === 'undefined') {
            this.client.logger.error(`[GROUP-${this.id}] Server ${serverId} not found in group. Attempting to re-add server to group.`);
            server = await this.addServer(serverId);
        }
        return server;
    }
    async dispose() {
        this.removeServers();
        await Promise.all([
            this.client.subscriptions.unsubscribe('group-update', this.id.toString()),
            this.client.subscriptions.unsubscribe('group-server-create', this.id.toString()),
            this.client.subscriptions.unsubscribe('group-server-delete', this.id.toString()),
            this.client.subscriptions.unsubscribe('group-server-status', this.id.toString()),
            this.client.subscriptions.unsubscribe('group-server-heartbeat', this.id.toString()),
            this.client.subscriptions.unsubscribe('group-member-update', this.id.toString())
        ]);
    }
}
